src/main/java/hudson/security/UserAttributesHelper.java

```diff
@@ -0,0 +1,190 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.security;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.naming.NamingException;
+import javax.naming.directory.Attribute;
+import javax.naming.directory.Attributes;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.GregorianCalendar;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Ease all the computations required to determine the user account optional attributes for creating
+ * the UserDetails that will be used by the SecurityRealm
+ */
+@Restricted(NoExternalUse.class)
+public class UserAttributesHelper {
+    // https://support.microsoft.com/en-us/help/305144/how-to-use-the-useraccountcontrol-flags-to-manipulate-user-account-pro
+    private static String ATTR_USER_ACCOUNT_CONTROL = "userAccountControl";
+    private static String ATTR_ACCOUNT_EXPIRES = "accountExpires";
+    // for Windows Server 2003-based domain
+    private static String ATTR_USER_ACCOUNT_CONTROL_COMPUTED = "msDS-User-Account-Control-Computed";
+    // for ADAM (Active Directory Application Mode), replace the ADS_UF_DISABLED
+    private static String ATTR_USER_ACCOUNT_DISABLED = "msDS-UserAccountDisabled";
+    // for ADAM, replace the ADS_UF_PASSWORD_EXPIRED
+    private static String ATTR_USER_PASSWORD_EXPIRED = "msDS-UserPasswordExpired";
+
+    // https://docs.microsoft.com/en-us/windows/desktop/adschema/a-accountexpires
+    // constant names follow the code in Iads.h
+    private static long ACCOUNT_NO_EXPIRATION = 0x7FFFFFFFFFFFFFFFL;
+    private static int ADS_UF_DISABLED = 0x0002;
+    private static int ADS_UF_LOCK_OUT = 0x0010;
+    private static int ADS_UF_PASSWORD_EXPIRED = 0x800000;
+
+    public static boolean checkIfUserIsEnabled(@Nonnull Attributes user) {
+        try {
+            String userAccountControl = getStringAttribute(user, ATTR_USER_ACCOUNT_CONTROL);
+            if (userAccountControl != null) {
+                int uacAsInt = Integer.parseInt(userAccountControl);
+                if ((uacAsInt & ADS_UF_DISABLED) == ADS_UF_DISABLED) {
+                    return false;
+                }
+            }
+
+            String adamUserAccountDisabled = getStringAttribute(user, ATTR_USER_ACCOUNT_DISABLED);
```

Comment:
Is this a typo? adam?

Reply:
Active directory application mode, cf comment in the `ATTR_USER_ACCOUNT_DISABLED` field. It's a particular mode for AD, not sure exactly about the details, but the standard is different for that mode.

Reply:
Oh, ok, thanks!

Reply:
After a second reading...

> standard is different

That's exactly the root cause of my non-confidence in those PRs, the lack of standard... :D